protocol/lavaprotocol/finalization_consensus.go

```diff
@@ -100,83 +99,113 @@ func (fc *FinalizationConsensus) insertProviderToConsensus(blockDistanceForFinal
-		// go over all consensus groups, if there is a mismatch add it as a consensus group and send a conflict
-		for _, consensus := range fc.currentProviderHashesConsensus {
-			err := fc.discrepancyChecker(finalizedBlocks, consensus)
-			if err != nil {
-				// TODO: bring the other data as proof
-				finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: reply}
-				// we need to insert into a new consensus group before returning
-				// or create new consensus group if no consensus matched
-				continue
-			}
-
-			if !inserted {
-				// if no discrepency with this group and not inserted yet -> insert into consensus
-				fc.insertProviderToConsensus(blockDistanceForFinalizedData, &consensus, finalizedBlocks, latestBlock, reply, req, providerAddress)
-				inserted = true
-			}
-			// keep comparing with other groups, if there is a new message with a conflict we need to report it too
-		}
-		if !inserted {
-			// means there was a consensus mismatch with everything, so it wasn't inserted and we add it here
-			newHashConsensus := fc.newProviderHashesConsensus(blockDistanceForFinalizedData, providerAddress, latestBlock, finalizedBlocks, reply, req)
-			fc.currentProviderHashesConsensus = append(fc.currentProviderHashesConsensus, newHashConsensus)
-		}
-		if finalizationConflict != nil {
-			// means there was a conflict and we need to report
-			return finalizationConflict, utils.LavaFormatError("Simulation: Conflict found in discrepancyChecker", err)
+		// Could not find discrepancy, let's check with previous epoch
+		foundDiscrepancy, discrepancyBlock = fc.findDiscrepancy(finalizedBlocks, fc.prevEpochBlockToHashesToAgreeingProviders)
+		if foundDiscrepancy {
+			blockToHashesToAgreeingProviders = fc.prevEpochBlockToHashesToAgreeingProviders
+		} else {
+			// No discrepancy found, log and return nil
+			logSuccessUpdate()
+			return finalizationConflict, nil
 		}
+	}
+
+	replyFinalization := conflicttypes.NewRelayFinalizationMetaDataFromRelaySessionAndRelayReply(relaySession, reply, consumerAddress)
+	finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: &replyFinalization}
 
-		// check for discrepancy with old epoch
-		for idx, consensus := range fc.prevEpochProviderHashesConsensus {
-			err := fc.discrepancyChecker(finalizedBlocks, consensus)
-			if err != nil {
-				// TODO: bring the other data as proof
-				finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: reply}
-				return finalizationConflict, utils.LavaFormatError("Simulation: prev epoch Conflict found in discrepancyChecker", err, utils.Attribute{Key: "Consensus idx", Value: strconv.Itoa(idx)}, utils.Attribute{Key: "provider", Value: providerAddress})
+	var otherBlockHash string
+	for blockHash, agreeingProviders := range blockToHashesToAgreeingProviders[discrepancyBlock] {
+		if providerPrevReply, ok := agreeingProviders[providerAddress]; ok {
+			// Same provider conflict found - Report and block provider
```

Comment:
There are 2 cases here: 
- A provider is lying about the hashes, but it's consistent, meaning - the hashes are consistent between requests for the same provider. In that case, there won't be Same Provider Conflict, because the provider is not conflicting with itself, but, there will probably be Two Providers Finalization Conflict, since this provider and other providers won't be "agreeingProviders', and they will agree on different hashes.
- A provider is lying about the hashes, but it's inconsistent, meaning - first time asked, it will return some hash on block 100, second time asked, it will return different hash for the same 100th block. In this case, the same provider will under "agreeingProviders" of two or more different hashes, which is covered in the lines above, since I'm looping over all the hashes recorded for each block. In this case, there will be a Same Provider Conflict. 

I hope that this clarifies it a little bit.